README.rst
========================================
Python library and CLI for UW Groups API
========================================

.. contents:: Table of Contents

* API Documentation: https://nhoffman.github.io/uwgroups/

Implements a limited subset of the UW Groups REST API v3
(https://wiki.cac.washington.edu/display/infra/Groups+Service+API+v3)
that I find useful in a specific application.

dependencies
============

* Python 3.6+
* Tested on Linux and OS X.

installation
============

Set up a virtualenv::

  python3 -m venv py3-env
  source py3-env/bin/activate
  pip install -U pip

For now, install directly from GitHub::

  pip install git+https://github.com/nhoffman/uwgroups.git

Alternatively, for development, from the top level of the repo (after
installing and activating the virtualenv)::

  pip install -e .
  pip install -r requirements.txt

CLI
===

After installation, the CLI entry point is called ``uwgroups``. For
development, you can invoke the ``uwgroups.py`` from within the cloned
repo::

    % cd uwgroups
    % ./uwgroups.py --help

or::

   % path/to/uwgroups/uwgroups.py --help

When invoked this way, the local version of the package is imported,
even if the version of the package is installed to the system. This is
very handy for development, and can avoid the requirement for a
virtualenv in many cases.

Commands are constructed as follows. Every command starts with the
name of the script, followed by an "action" followed by a series of
required or optional "arguments". The name of the script, the action,
and options and their arguments are entered on the command line
separated by spaces. Help text is available for both the ``uwgroups``
script and individual actions using the ``-h`` or ``--help`` options.

authentication
==============

Authentication to the UW Groups API uses SSL/TLS authentication with
an X.509 certificate in pem format along with the private key used to
generate it. The certificate and key may be concatenated into the same
file, for example::

  -----BEGIN RSA PRIVATE KEY-----
  <key contents>
  -----END RSA PRIVATE KEY-----
  -----BEGIN CERTIFICATE-----
  <cert contents>
  -----END CERTIFICATE-----

In this case, the concatenated key and cert can either be specified on
the command line using ``-c/--cert-file`` before the name of the
subcommand::

  uwgroups -c certfile.pem connect

or using the environment variable ``UWGROUPS_CERT``::

  export UWGROUPS_CERT=certfile.pem
  uwgroups connect

The key may also be provided in a separate fie using ``-k/--key-file``
or ``UWGROUPS_KEY`` as above.

API hosts
=========

Use the environment variable ``GWS_HOST`` to define the API host:

* ``PROD``: groups.uw.edu (default)
* ``DEV``: dev.groups.uw.edu
* ``EVAL``: eval.groups.uw.edu

unit tests
==========

Unit tests are implemented using the ``unittest`` module in the Python
standard library. The ``tests`` subdirectory is itself a Python
package that implements the tests. All unit tests can be run like this::

    % python setup.py test

A single unit test can be run by referring to a specific module,
class, or method within the ``tests`` package using dot notation::

    % python setup.py test --test-suite tests.test_utils

license
=======

Copyright (c) 2017-2020 Noah Hoffman

Released under the MIT License:

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGES.rst
======================
 changes for uwgroups
======================

0.3.7
=====

* add UWGroups.search_group() to return a list of group names based on
  a pattern.

0.3.6
=====

* Work around error caused by 1024 bit UWCA root cert
  (ssl.SSLError: [SSL: CA_KEY_TOO_SMALL] ca key too small (_ssl.c:2633))

0.3.5
=====

* add batchsize parameter to UWGroups.sync_members()
* change default batchsize to 50

0.3.4
=====

* add subcommands to add and remove users from groups
* recursively create parent groups on group creation

0.3.3
=====

* fix error parsing output of openssl

0.3.2
=====

* add 'UWGroups.search_user()' method and corresponding subcommand

0.3.1
=====

* add 'environment' parameter to UWGroups() constructor

0.3.0
=====

* drop python 2.7, support python 3.6+
* use UW groups API v3
  (https://wiki.cac.washington.edu/display/infra/Groups+Service+API+v3)
* use environment variable GWS_HOST to specify API host
* add subcommand 'members'

0.2.5
=====

* Use port 443 instead of 7443

0.2.4
=====

* Better handling of timeout errors.

0.2.3
=====

* add ``timeout`` parameter to ``UWGroups`` constructor and
  ``UWGroups.connect()`` with default of 30s



uwgroups/api.py
import errno
import logging
import http.client
from os import path
import socket
import subprocess
import json

from uwgroups import package_data
from uwgroups.utils import reconcile, grouper, check_types

log = logging.getLogger(__name__)

# https://groups.uw.edu/group_sws/v3
GWS_HOSTS = {
    'PROD': 'groups.uw.edu',
    'DEV': 'dev.groups.uw.edu',
    'EVAL': 'eval.groups.uw.edu',
}

GWS_PORT = 443
API_PATH = '/group_sws/v3'

# https://certs.cac.washington.edu/?req=svpem
UWCA_ROOT = package_data('root.cert')


class APIError(Exception):
    pass


class MissingResourceError(APIError):
    pass


class AuthorizationError(APIError):
    pass


def get_admins(certfile):
    """Returns [dns_user, uwnetid_user] given data in cert

    """

    output = subprocess.run(
        ['openssl', 'x509', '-in', certfile, '-noout', '-subject'],
        capture_output=True, text=True)
    log.info(output)

    # openssl and libressl have differently-formatted outputs:
    # LibreSSL 2.8.3
    # subject= /C=US/ST=Washington/L=Seattle/O=University of Washington/OU=Laboratory Medicine/CN=elmira.labmed.uw.edu/emailAddress=ngh2@uw.edu
    # openssl
    # subject=C = US, ST = Washington, L = Seattle, O = University of Washington, OU = Laboratory Medicine, CN = elmira.labmed.uw.edu, emailAddress = ngh2@uw.edu

    text = output.stdout.replace('subject=', '').strip()
    char = '/' if text.startswith('/') else ','
    data = {k.strip(): v.strip()
            for k, v in [e.split('=') for e in text.strip(char).split(char)]}

    dns_user = User(uwnetid=data['CN'], type='dns')
    uwnetid_user = User(
        uwnetid=data['emailAddress'].split('@')[0], type='uwnetid')

    return [dns_user, uwnetid_user]


class User(object):
    user_types = {'uwnetid', 'group', 'dns', 'eppn'}

    def __init__(self, uwnetid, type):
        if type not in self.user_types:
            raise ValueError('user_type must be in {}'.format(self.user_types))

        self.uwnetid = uwnetid
        self.type = type

    def __repr__(self):
        return 'User(uwnetid={}, type={})'.format(self.uwnetid, self.type)


class UWGroups(object):
    """Class providing a connection to the UW groups REST
    API. ``certfile`` and ``keyfile`` are paths to files containing
    the certificate and private keys, respectively. ``timeout`` is
    passed to ``httplib.HTTPSConnection()``

    As of 2022-06-28, the UWCA root cert is only 1024 bytes, which
    results in the error

    'ssl.SSLError: [SSL: CA_KEY_TOO_SMALL] ca key too small (_ssl.c:2633)'
    on ubuntu and debian. By default, the connection is initialized with
    context.set_ciphers('DEFAULT@SECLEVEL=1') unless use_default_ciphers is True

    Example::

        with UWGroups(certfile='/path/to/cert.pem') as conn:
            conn.get_group('some_group')

    """

    def __init__(self, certfile, keyfile=None, environment='PROD', timeout=30,
                 use_default_ciphers=False):
        """Initialize the connection.

        """

        self.gws_host = GWS_HOSTS[environment]
        log.info(f'using {self.gws_host}')

        if not certfile or not path.exists(certfile):
            raise ValueError(
                "'certfile' is required and must specify a readable file")

        if keyfile and not path.exists(keyfile):
            raise ValueError("'keyfile' must specify a readable file")

        self.keyfile = keyfile
        self.certfile = certfile
        self.admins = get_admins(certfile)
        log.info(self.admins)
        self.timeout = timeout
        self.use_default_ciphers = use_default_ciphers
        self.context = None
        self.connection = None

    def connect(self, timeout=None):
        """Establish a connection. If the ``UWGroups`` object is instantiated
        without a with block, must be called explicitly. ``timeout``
        overrides the value for ``timeout`` in the class constructor.
        """
        self.context = http.client.ssl.create_default_context()
        self.context.load_cert_chain(certfile=self.certfile, keyfile=self.keyfile)
        self.context.load_verify_locations(UWCA_ROOT)
        if not self.use_default_ciphers:
            self.context.set_ciphers('DEFAULT@SECLEVEL=1')
        self.connection = http.client.HTTPSConnection(
            host=self.gws_host,
            timeout=timeout or self.timeout,
            port=GWS_PORT,
            context=self.context)
        log.info(f'connected to {self.gws_host}:{GWS_PORT}')

    def __enter__(self):
        self.connect()
        return self

    def __exit__(self, type, value, tback):
        self.close()

    def close(self):
        """Close the connection"""
        self.connection.close()

    def reset(self):
        """Close and reopen the connection; may be necessary after an
        exception"""
        self.close()
        self.connect()

    @check_types(method=str, endpoint=str, headers=dict,
                 body=str, expect_status=int, attempts=int)
    def _request(self, method, endpoint, headers=None, body=None,
                 expect_status=200, attempts=5):
        methods = {'GET', 'PUT', 'DELETE'}
        if method not in methods:
            raise ValueError(
                'method must be one of {}'.format(', '.join(methods)))

        url = path.join(API_PATH, endpoint)

        args = {}
        if headers:
            args['headers'] = headers
        if body:
            args['body'] = body

        for attempt in range(attempts):
            try:
                self.connection.request(method, url, **args)
                response = self.connection.getresponse()
            except http.client.BadStatusLine as err:
                caught_err = err
                log.warning('failure on attempt {}: {}'.format(attempt, err))
                self.reset()
            except socket.error as err:
                caught_err = err
                if err.errno != errno.ETIMEDOUT:
                    log.warning('failure on attempt {}: {}'.format(attempt, err))
                    # response isn't set at all for this case, reraise
                    # the exception.
                    raise
                self.reset()
            else:
                break
        else:
            # Hit the limit of attempts. The response variable never got set.
            msg = '{}: {} {}'.format(caught_err, method, url)
            log.info(msg)
            raise APIError(msg)

        msg = '{} {}: {} {}'.format(method, url, response.status, response.reason)
        log.info(msg)

        if response.status == 404:
            raise MissingResourceError(msg)
        elif response.status == 401:
            raise AuthorizationError(msg)
        elif response.status != expect_status:
            raise APIError(msg)

        body = response.read()
        return body

    @check_types(group_name=str)
    def group_exists(self, group_name):
        """Return True if the group exists"""
        try:
            self.get_group(group_name)
        except MissingResourceError:
            # closing and reopening the connection seems to be
            # necessary to prevent an httplib.ResponseNotReady error
            # after an exception caused by a missing group.
            self.reset()
            return False
        else:
            return True

    @check_types(group_name=str)
    def get_group(self, group_name):
        """Return deserialized json representation of a group"""
        endpoint = path.join('group', group_name)
        response = self._request(
            'GET', endpoint,
            headers={"Accept": "application/json",
                     "Content-Type": "application/json"})
        return json.loads(response)

    @check_types(group_name=str, admin_users=list)
    def create_group(self, group_name, admin_users=None):
        """Create a group with name ``group_name``. The dns user
        associated with the certificate, as well as the user
        identified as the email contact for the cert in field
        ``emailAddress`` are always added as admins. A list of
        additional admin users identified by uwnetid can be provided
        in ``admin_users``.

        Note that the API requires that a parent group ('root_parent')
        must be created before any of its children
        ('root_parent_child'). This method will recursively create
        parent groups if necessary.

        """

        if self.group_exists(group_name):
            log.info(f'group {group_name} exists, skipping creation')
            return {}

        # create parents as necessary, skipping the first (root) group
        hierarchy = group_name.split('_')
        for i in range(2, len(hierarchy)):
            parent = '_'.join(hierarchy[:i])
            self.create_group(parent, admin_users=admin_users or [])

        endpoint = path.join('group', group_name)

        admins = self.admins[:]
        if admin_users:
            for uwnetid in admin_users:
                admins.append(User(uwnetid, type='uwnetid'))

        body = {
            "data": {
                "id": group_name,
                "displayName": group_name,
                "description": group_name,
                "contact": "ngh2",
                "authnfactor": 1,
                "classification": "u",
                "admins": [{'id': u.uwnetid, 'type': u.type} for u in admins],
                "updaters": [],
                "creators": [],
                "readers": [],
                "optins": [],
                "optouts": [],
            }
        }

        response = self._request(
            'PUT', endpoint,
            headers={"Accept": "application/json",
                     "Content-Type": "application/json"},
            body=json.dumps(body),
            expect_status=201)
        log.debug(response)
        return response

    @check_types(group_name=str)
    def delete_group(self, group_name):
        endpoint = path.join('group', group_name)
        response = self._request('DELETE', endpoint)
        return response

    @check_types(group_name=str)
    def get_members(self, group_name):
        """Return a list of uwnetids"""
        endpoint = path.join('group', group_name, 'member')
        response = self._request(
            'GET', endpoint, headers={'accept': 'application/json'})
        data = json.loads(response)['data']
        members = [d['id'] for d in data]
        return members

    @check_types(group_name=str, members=list, batchsize=int)
    def add_members(self, group_name, members, batchsize=50):
        """Add uwnetids in list ``members`` to the specified group in batches
        of size ``batchsize``.

        """
        for chunk in grouper(members, batchsize):
            endpoint = path.join('group', group_name, 'member', ','.join(chunk))
            self._request('PUT', endpoint)

    @check_types(group_name=str, members=list, batchsize=int)
    def delete_members(self, group_name, members, batchsize=50):
        """Remove uwnetids in list ``members`` from the specified group in
        batches of size ``batchsize``.

        """
        for chunk in grouper(members, batchsize):
            endpoint = path.join('group', group_name, 'member', ','.join(chunk))
            self._request('DELETE', endpoint)

    @check_types(group_name=str, members=list)
    def sync_members(self, group_name, members, batchsize=50, dry_run=False):
        """Add or remove users from the specified group as necessary so that
        the group contains ``members`` (a list or uwnetids). When
        ``dry_run`` is True, log the necessary actions but don't
        modify the group. The group is created if it does not already
        exist.

        """

        desired_members = set(members)

        try:
            current_members = set(self.get_members(group_name))
        except MissingResourceError:
            self.reset()
            current_members = set()
            log.warning('creating group {}'.format(group_name))
            if not dry_run:
                self.create_group(group_name)

        log.info('{} current members'.format(len(current_members)))
        log.debug('current members: {}'.format(current_members))
        to_add, to_delete = reconcile(current_members, desired_members)

        if to_add:
            log.info('[+] {}: {}'.format(group_name, ','.join(to_add)))
            if not dry_run:
                self.add_members(group_name, sorted(to_add), batchsize=batchsize)

        if to_delete:
            log.info('[-] {}: {}'.format(group_name, ','.join(to_delete)))
            if not dry_run:
                self.delete_members(
                    group_name, sorted(to_delete), batchsize=batchsize)

    @check_types(group_name=str, service=str, active=bool)
    def set_affiliate(self, group_name, service, active=True):
        """Activate (``active=True``) or inactivate (``active=False``)
        Exchange (``service=exchange``) or Google Apps
        (``service=google``) for the specified group.

        """

        services = {'exchange': 'email', 'google': 'google'}
        if service not in services:
            raise ValueError('service must be one of {}'.format(
                list(services.keys())))
        endpoint = path.join('group', group_name, 'affiliate', services[service])
        endpoint += '?status=' + ('active' if active else 'inactive')
        response = self._request('PUT', endpoint)
        return response

    @check_types(netid=str)
    def search_user(self, netid):
        """Return groups in which `netid` is a member

        """

        endpoint = f'search?member={netid}'
        response = self._request('GET', endpoint)
        data = json.loads(response)['data']
        groups = [d['id'] for d in data]
        return groups

    @check_types(name=str)
    def search_groups(self, name):
        """Return a list of group names matching pattern `name` (may
        contain * as a wildcard).

        """

        endpoint = f'search?name={name}'
        response = self._request('GET', endpoint)
        data = json.loads(response)['data']
        groups = [d['id'] for d in data]
        return groups
